### PR TITLE
rgw: address bug where object puts could write to decommissioned shard

### PR DESCRIPTION
**Pull request by ivancich**
Addresses and an apparent bug where some puts that initiate around the time that resharding completes are written to the old bucket shard rather than the new one.

Signed-off-by: J. Eric Ivancich [ivancich@redhat.com](mailto:ivancich@redhat.com)

Fixes: https://tracker.ceph.com/issues/58034

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by ivancich** [comment]
@mkogan1, would you be able to run your same test on main (not necessarily this branch as it's only additional commit is one with vstart config changes for my own testing). Where I was able to reproduce the issue with pacific and your test, Mark, I cannot reproduce it on main. Wondering if it was fixed somehow else. Thanks!

**Comment by cbodley** [comment]
@ivancich it looks like this branch is missing a commit. i only see the TEMPORARY one

**Comment by ivancich** [comment]
> @ivancich it looks like this branch is missing a commit. i only see the TEMPORARY one

That's correct. It was created before code ported over to start conversation with Mark Kogan.

**Comment by ivancich** [comment]
Teuthology Results -- no regressions:
* http://pulpito.ceph.com/ivancich-2022-11-29_20:34:44-rgw-wip-fix-old-shard-distro-default-smithi/
* http://pulpito.ceph.com/ivancich-2022-11-29_20:35:09-rgw-wip-fix-old-shard-distro-default-smithi/

**Comment by ivancich** [comment]
jenkins test make check

**Comment by ivancich** [comment]
@cbodley : This passed teuthology testing. When you get a chance, could you review?

**Review by cbodley — APPROVED** [review]
looks great!

File: src/cls/rgw/cls_rgw_types.h [hunk 1]
```diff
@@ -845,9 +846,15 @@ struct cls_rgw_bucket_instance_entry {
...
   }
+
   bool resharding_in_progress() const {
     return reshard_status == RESHARD_STATUS::IN_PROGRESS;
   }
+
+  friend std::ostream& operator<<(std::ostream& out, const cls_rgw_bucket_instance_entry& v) {
+    out << "instance entry reshard status: " << v.reshard_status;
+    return out;
+  }
```
[the remark this answers is not in the document]
Comment (ivancich) [reply src/cls/rgw/cls_rgw_types.h:857]: Probably. I added it during development. I don't see a harm in leaving it in, so it's there for future development. Do you?

**Pull request merged**